tests/0098-consumer-txn.cpp

```diff
@@ -0,0 +1,851 @@
+static int64_t partition_0_hi_offset;
+static int64_t partition_0_ls_offset;
+
+class TestEventCb : public RdKafka::EventCb {
+ public:
+  void event_cb (RdKafka::Event &event) {
+
+    switch (event.type())
+    {
+      case RdKafka::Event::EVENT_STATS:
+        if (_should_capture_stats) {
+          _has_captured_stats = true;
+          _should_capture_stats = false;
+
+          std::vector<std::string> hi_path;
+          hi_path.push_back("\"partitions\"");
+          hi_path.push_back("\"0\"");
+          hi_path.push_back("\"hi_offset\"");
+          partition_0_hi_offset = extract_json_value(event.str(), hi_path);
+
+          std::vector<std::string> ls_path;
+          ls_path.push_back("\"partitions\"");
+          ls_path.push_back("\"0\"");
+          ls_path.push_back("\"ls_offset\"");
+          partition_0_ls_offset = extract_json_value(event.str(), ls_path);
+        }
+        break;
+
+      default:
+        break;
+    }
+  }
+};
+
+static TestEventCb ex_event_cb;
+
+
+static void test_assert(bool cond, std::string msg) {
+  if (!cond)
+    Test::Say(msg);
+  assert(cond);
+}
+
+
+static void execute_java_produce_cli(std::string &bootstrapServers,
+                                     std::string &topic, std::string cmd) {
+  std::stringstream ss;
+  ss << "./java/run-class.sh TransactionProducerCli " + 
+        bootstrapServers + " " + topic + " " + cmd;
+  int status = system(ss.str().c_str());
+  test_assert(!status, 
+              tostr() << "./java/run-class.sh TransactionProducerCli failed "
+                         "with error code: "
+                      << status);
+}
+
+static std::vector<RdKafka::Message *> consume_messages(
+                                          RdKafka::KafkaConsumer *c, 
+                                          std::string topic,
+                                          int partition) {
+  RdKafka::ErrorCode err;
+  int32_t limit_count;
+
+  /* Assign partitions */
+  std::vector<RdKafka::TopicPartition*> parts;
+  parts.push_back(RdKafka::TopicPartition::create(topic, partition));
+  if ((err = c->assign(parts)))
+    Test::Fail("assign failed: " + RdKafka::err2str(err));
+  RdKafka::TopicPartition::destroy(parts);
+
+  Test::Say("Consuming from topic " + topic + "\n");  
+  std::vector<RdKafka::Message *> result = std::vector<RdKafka::Message *>();
+
+  while (true) {
+    RdKafka::Message *msg = c->consume(tmout_multip(1000));
+    switch (msg->err())
+    {
+      case RdKafka::ERR__TIMED_OUT:
+        continue;
+      case RdKafka::ERR__PARTITION_EOF:
+        break;
+      case RdKafka::ERR_NO_ERROR:
+        result.push_back(msg);
+        continue;
+      default:
+        Test::Fail("Error consuming from topic " + 
+                   topic + ": " + msg->errstr());
+        break;
+    }
+    break;
+  }
+
+  Test::Say("Read all messages from topic: " + topic + "\n");
+
+  _should_capture_stats = true;
+  limit_count = 0;
+  while (limit_count++ < 20) {
```

Comment:
Where's the value 20 coming from?

Reply:
yes, this is stupid, i just need to block the consume longer than the stats interval.

Reply:
Then check time and not iterations, if consume() wakes up from other events it makes the test unreliable.

You can even rely on the test timeout to take care of it for you, just wait indefinitely for the break event, the test timeout will kill the test if it runs for longer than the timeout you set in test_conf_init

Reply:
using # iterations is dead simple and didn't involve me doing an additional 2 minutes of irksome investigation into how to check time in C, and/or if there is some other idiomatic way you do it in librdkafka (did i mention i need a vacation?).

anyway, relying on test timeout is a great idea, going with that.